File: backend_python/src/todo_api/application_service/usecase/get_todo_usecase.py
from abc import ABC, abstractmethod

from pydantic import BaseModel, ConfigDict

from todo_api.domain.model.todo import Todo
from todo_api.domain.repository.todo_repository import TodoRepository
from todo_api.utils.uuid import UUID7


class GetTodoUsecaseInput(BaseModel):
    model_config = ConfigDict(arbitrary_types_allowed=True)
    id: UUID7


class GetTodoUsecaseOutput(BaseModel):
    model_config = ConfigDict(arbitrary_types_allowed=True)

    todo: Todo


class GetTodoUsecase(ABC):
    @abstractmethod
    def execute(self, input_dto: GetTodoUsecaseInput) -> GetTodoUsecaseOutput:
        pass


class GetTodoUsecaseImpl(GetTodoUsecase):
    def __init__(self, todo_repository: TodoRepository) -> None:
        self.todo_repository = todo_repository

    def execute(self, input_dto: GetTodoUsecaseInput) -> GetTodoUsecaseOutput:
        todo = self.todo_repository.find_by_id(input_dto.id)
        return GetTodoUsecaseOutput(todo=todo)

File: backend_bun/src/application_service/usecase/getTodoUsecase.ts
import type { Todo } from "@/domain/model/todo";
import type { TodoRepository } from "@/domain/repository/todoRepository";

export type GetTodoRequest = {
  id: string;
};

export type GetTodoResponse = {
  todo: Todo;
};

export interface GetTodoUsecase {
  execute(request: GetTodoRequest): Promise<GetTodoResponse>;
}

export class GetTodoUsecaseImpl implements GetTodoUsecase {
  constructor(private todoRepository: TodoRepository) {}

  async execute(request: GetTodoRequest): Promise<GetTodoResponse> {
    const todo = await this.todoRepository.findById(request.id);
    return { todo };
  }
}
